server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -782,6 +782,11 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
             o.writeInt(period.getYears());
             o.writeInt(period.getMonths());
             o.writeInt(period.getDays());
+        }),
+        entry(GenericNamedWriteable.class, (o, v) -> {
+            o.writeByte((byte) 30);
+            GenericNamedWriteable w = (GenericNamedWriteable) v;
+            o.writeNamedWriteable(w);
         })
```

Comment:
This is the extra validation I think we should have:

```suggestion
        entry(GenericNamedWriteable.class, (o, v) -> {
            final var genericNamedWriteable = (GenericNamedWriteable) v;
            final var minSupportedVersion = TransportVersion.max(
                TransportVersion.V_8_500_069,
                genericNamedWriteable.getMinimalSupportedVersion()
            );

            if (o.getTransportVersion().before(minSupportedVersion)) {
                final var message = Strings.format(
                    "[%s] requires minimal transport version [%s] and cannot be sent using transport version [%s]",
                    genericNamedWriteable,
                    minSupportedVersion,
                    o.getTransportVersion()
                );
                assert false : message;
                throw new IllegalStateException(message);
            }

            o.writeByte((byte) 30);
            o.writeNamedWriteable(genericNamedWriteable);
        })
```

Reply:
That code looks good, but does it really do anything different to https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/io/stream/VersionCheckingStreamOutput.java#L64? That code should be called already, inside the call to `writeNamedWritable`, so the check already exists, as far as I can tell.

Reply:
What I would like would be compatibility checks that verify that the exception is indeed thrown when we have mix-clusters. I wrote some tests like this before (in other work), but they were flaky (throwing the expected exception 99% of the time, not 100%, which was odd). I can try again.

Reply:
`VersionCheckingStreamOutput` doesn't consider it a bug to try and send an unknown writable, it just throws an IAE which maps onto a `400 Bad Request` and isn't logged. I think that's unduly lenient (although I see why checking this properly is tricky in a cross-cluster situation) and I'd rather we were stricter where possible, asserting that this doesn't happen and throwing an ISE if ever it does.

Reply:
> What I would like would be compatibility checks that verify that the exception is indeed thrown when we have mix-clusters.

I think a unit test is enough for this. We shouldn't even be trying to send unknown stuff over the wire ideally, hence why I think we should assert that we don't. Instead I'd rather that we looked at `org.elasticsearch.cluster.ClusterState#getMinTransportVersion` and avoided even creating the problematic object until all nodes can understand it.

Reply:
I'm thinking, if we have a mixed cluster, and there is a query involving a new feature that requires sending something new over the wire, that only some of the cluster understands, the consequence is:

* If the sender is an old version, it will not even get to this point because the new feature is not understood causing an exception in the feature itself (eg. in my case the painless script will fail with a painless exception).
* If the sender is new, but the receiver is old, then the sender serialises type=30, and the receiver throws `IOException("Can't read unknown type [" + type + "]")` in `StreamInput.readGenericValue`.

So, as far as I can tell your suggestion is that instead of the receiver throwing this exception, we have a sender that detects the receiver is old and throws an exception before sending, which moves the error to a different node in the cluster. Is this a big improvement?

In fact there are three scenarios:

* If we do no checking of versions at all (ie. my initial code) we get the behaviour described above
* If we check with VersionedNamedWritable (ie. my current code) we get an exception in the sender if the VersionedNamedWritable version is later than the senders TransportVersion which only considers local node state, so does not care if the remote node has an older version
* We check the remote transport version also, and throw an exception there. Looking at your code, it does not do this but seems to consider local transport version only, so very similar to the current behaviour. Perhaps your last suggestion of using `ClusterState#getMinTransportVersion` is the way to do this? But again, the consequence is only to move the exception to a different node in the cluster.

Reply:
I've done a little more digging and it looks like VersionedNamedWritable checks only occur for CCS, way too narrow a scope for our needs, so it is clear that I do need to write the check you have suggested. I'll go for that now.

Reply:
We should forbid the use of any features that the whole cluster does not understand (which is why I think we should be looking at `ClusterState#getMinTransportVersion`). At least, that's [the documented behaviour](https://www.elastic.co/guide/en/elastic-stack/8.9/upgrading-elasticsearch.html#rolling-upgrades). Then we should consider it a bug (and `assert false`) if ever we get into a position where we're trying to send a message that the remote won't understand.

> Looking at your code, it does not do this but seems to consider local transport version only

No, it's looking at `StreamOutput#getTransportVersion` which is a transport version compatible with both sender and receiver.

Reply:
I've investigated further and figured out why I misunderstood this. I did not see the uses of `setTransportVersion`, which allows for the behaviour you describe.